Core/Consumers/AbstractAsyncConsumer.php

```diff
@@ -0,0 +1,214 @@
+     */
+    public function consume(EndpointInterface $endpoint)
+    {
+        $this->initialize($endpoint);
+
+        $callback = $this->callback($endpoint);
+        $this->asyncConsume($endpoint, $callback);
+
+        while (!$this->shouldStop() && $this->sleep = true) {
+            try {
+                $this->waitNoBlock($endpoint);
+
+                if ($this->sleep) {
+                    usleep(self::SLEEP_DURATION);
+                }
+            } catch (\Exception $exception) {
+                if (!$this->stop) {
+                    throw $exception;
+                }
+            }
+        }
+
+        $this->cleanUp($endpoint);
+    }
+
+    /**
+     * Callback function to be triggered when a message is received. Is in charge of triggering the processing,
+     * confirming the message if no exception was thrown, reducing the expiration count and preventing the consumer
+     * from sleeping in the next round.
+     *
+     * @param EndpointInterface $endpoint
+     *
+     * @return \Closure
+     */
+    public function callback(EndpointInterface $endpoint): callable
+    {
+        return function (MessageInterface $message) use ($endpoint) {
+            $start = microtime(true);
+
+            try {
+                $this->process($endpoint, $message);
+            } catch (\Exception $exception) {
+                $this->consumptionDuration = (microtime(true) - $start) * 1000;
+                $this->dispatchConsumerTimingEvent($message);
+
+                throw $exception;
+            }
+
+            $this->consumptionDuration += (microtime(true) - $start) * 1000;
+            $this->dispatchConsumerTimingEvent($message);
+
+            $this->confirmMessage($endpoint, $message);
+            $this->logConsumeMessage();
+            --$this->expirationCount;
+            $this->consumptionDuration = 0;
+            $this->sleep = false;
+        };
+    }
+
+    /**
+     * Hooks the consumer with the source of information and passes the callback function to be called once a message
+     * is received.
+     *
+     * @param EndpointInterface $endpoint
+     * @param callable $callback
+     */
+    abstract public function asyncConsume(EndpointInterface $endpoint, callable $callback);
+
+    /**
+     * Waits for a message in a blocking way. If the worker needs to listen to signals, use waitNoBlock() instead. This
+     * function won't return the control to the consumer until a message arrives and the callback function finishes.
+     *
+     * @param EndpointInterface $endpoint
+     */
+    abstract public function wait(EndpointInterface $endpoint);
+
+    /**
+     * Waits for a message in a non-blocking way. If there's no message to consume, control is returned to the consumer.
+     * Needs to be called in a loop in order to keep checking for messages.
+     *
+     * @param EndpointInterface $endpoint
+     */
+    abstract public function waitNoBlock(EndpointInterface $endpoint);
```

Comment:
Is it needed the functions `wait` and `waitNoBlock` to be public?

Reply:
Functions protected + related tests removed (with the new visibility they became invalid)

Reply:
:+1: beautiful commit!